src/bun.js/webcore/body.zig

```diff
@@ -1258,3 +1263,345 @@ pub fn BodyMixin(comptime Type: type) type {
+        global: *JSGlobalObject,
+        allocator: std.mem.Allocator,
+    ) @This() {
+        const this = .{
+            .ctx = ctx,
+            .onFinishedBuffering = onFinish,
+            .allocator = allocator,
+            .global = global,
+            .stream_buffer = .{
+                .allocator = allocator,
+                .list = .{
+                    .items = &.{},
+                    .capacity = 0,
+                },
+            },
+        };
+        return this;
+    }
+
+    pub fn run(sink: *@This(), value: *JSC.WebCore.Body.Value) !void {
+        value.toBlobIfPossible();
+
+        switch (value.*) {
+            .Used => {
+                log("Used", .{});
+                return error.StreamAlreadyUsed;
+            },
+            .Empty, .Null => {
+                log("Empty", .{});
+                return sink.onFinishedBuffering(sink.ctx, "", null, false);
+            },
+
+            .Error => |err| {
+                log("Error", .{});
+                return sink.onFinishedBuffering(sink.ctx, "", err, false);
+            },
+            // .InlineBlob,
+            .WTFStringImpl,
+            .InternalBlob,
+            .Blob,
+            => {
+                // toBlobIfPossible checks for WTFString needing a conversion.
+                var input = value.useAsAnyBlobAllowNonUTF8String();
+                const is_pending = input.needsToReadFile();
+                defer if (!is_pending) input.detach();
+
+                if (is_pending) {
+                    input.Blob.doReadFileInternal(*@This(), sink, onFinishedLoadingFile, sink.global);
+                } else {
+                    const bytes = input.slice();
+                    log("Blob {}", .{bytes.len});
+                    sink.onFinishedBuffering(sink.ctx, bytes, null, false);
+                }
+                return;
+            },
+            .Locked => {
+                try sink.bufferLockedBodyValue(value);
+            },
+        }
+    }
+
+    fn onFinishedLoadingFile(sink: *@This(), bytes: JSC.WebCore.Blob.Store.ReadFile.ResultType) void {
+        switch (bytes) {
+            .err => |err| {
+                log("onFinishedLoadingFile Error", .{});
+                sink.onFinishedBuffering(sink.ctx, "", err.toErrorInstance(sink.global), true);
+                return;
+            },
+            .result => |data| {
+                log("onFinishedLoadingFile Data {}", .{data.buf.len});
+                sink.onFinishedBuffering(sink.ctx, data.buf, null, true);
+                if (data.is_temporary) {
+                    bun.default_allocator.free(bun.constStrToU8(data.buf));
+                }
+            },
+        }
+    }
+    fn onStreamPipe(sink: *@This(), stream: JSC.WebCore.StreamResult, allocator: std.mem.Allocator) void {
+        var stream_needs_deinit = stream == .owned or stream == .owned_and_done;
+
+        defer {
+            if (stream_needs_deinit) {
+                if (stream.isDone()) {
```

Comment:
we probably should add a check for which union type it is just to be sure

Reply:
fixed in https://github.com/oven-sh/bun/pull/4418/commits/7cd66913e79f7c27d13efeb30801dfcbbabe45e6